slam_toolbox/lib/karto_sdk/src/Mapper.cpp

```diff
@@ -3017,38 +3056,112 @@ namespace karto
+        contrib::ComputeChowLiuTreeApproximation(
+            elimination_clique, local_marginal_covariance_matrix);
+    // (7) Push tree edges to graph and solver (as constraints).
+    for (Edge<LocalizedRangeScan> * edge : chow_liu_tree_approximation) {
+      bool edge_added = m_pGraph->AddEdge(edge);
+      assert(edge_added);  // otherwise internal logic is broken
+      m_pScanOptimizer->AddConstraint(edge);
+    }
+    return true;
+  }
+
+  kt_bool Mapper::RemoveEdgeFromGraph(Edge<LocalizedRangeScan> * edge_to_remove)
+  {
+    Vertex<LocalizedRangeScan> * source = edge_to_remove->GetSource();
+    Vertex<LocalizedRangeScan> * target = edge_to_remove->GetTarget();
+    source->RemoveEdge(edge_to_remove);
+    target->RemoveEdge(edge_to_remove);
+    m_pScanOptimizer->RemoveConstraint(
+        source->GetObject()->GetUniqueId(),
+        target->GetObject()->GetUniqueId());
+    m_pGraph->RemoveEdge(edge_to_remove);
+    delete edge_to_remove;
+    return true;
+  }
+
   kt_bool Mapper::RemoveNodeFromGraph(Vertex<LocalizedRangeScan>* vertex_to_remove)
   {
     // 1) delete edges in adjacent vertices, graph, and optimizer
-    std::vector<Vertex<LocalizedRangeScan>*> adjVerts =
-      vertex_to_remove->GetAdjacentVertices();
-    for (int i = 0; i != adjVerts.size(); i++)
-    {
-      std::vector<Edge<LocalizedRangeScan>*> adjEdges = adjVerts[i]->GetEdges();
+    std::vector<Vertex<LocalizedRangeScan> *> vertices =
+        vertex_to_remove->GetAdjacentVertices();
+    for (Vertex<LocalizedRangeScan> * vertex : vertices) {
       bool found = false;
-      for (int j=0; j!=adjEdges.size(); j++)
-      {
-        if (adjEdges[j]->GetTarget() == vertex_to_remove ||
-          adjEdges[j]->GetSource() == vertex_to_remove)
+      for (Edge<LocalizedRangeScan> * edge : vertex->GetEdges()) {
+        if (edge->GetTarget() == vertex_to_remove ||
+            edge->GetSource() == vertex_to_remove)
         {
-          adjVerts[i]->RemoveEdge(j);
+          vertex->RemoveEdge(edge);
           m_pScanOptimizer->RemoveConstraint(
-            adjEdges[j]->GetSource()->GetObject()->GetUniqueId(),
-            adjEdges[j]->GetTarget()->GetObject()->GetUniqueId()); 
-          std::vector<Edge<LocalizedRangeScan>*> edges = m_pGraph->GetEdges();
-          std::vector<Edge<LocalizedRangeScan>*>::iterator edgeGraphIt =
-            std::find(edges.begin(), edges.end(), adjEdges[j]);
-
-          if (edgeGraphIt == edges.end())
-          {
-            std::cout << "Edge not found in graph to remove!" << std::endl;
-            continue;
-          }
-
-          int posEdge = edgeGraphIt - edges.begin();
-          m_pGraph->RemoveEdge(posEdge); // remove from graph
-          delete *edgeGraphIt; // free hat!
-          *edgeGraphIt = NULL;
+              edge->GetSource()->GetObject()->GetUniqueId(),
+              edge->GetTarget()->GetObject()->GetUniqueId());
+          m_pGraph->RemoveEdge(edge);
+          delete edge;
```

Comment:
Need to set to null so that any future check for `if(ptr)` will return `false`, that will not be done if not set to null ptr

Reply:
Hmm, there are no future checks, and `edge` is scoped to the loop. Do you intend to set it to `nullptr` just to play it safe?

Reply:
Yeah, and there where places where this bit me in the ass badly when I was adding graph object removal and serialization. Anything deleted manually needs to have nullptr set in case something changes later. It needs to be possible to tell if a pointer is deleted or not. The only way is setting it to nullptr, just deleting it leaves it in limbo where `if(ptr)` checks don't return false.

Reply:
Alright. Done in https://github.com/SteveMacenski/slam_toolbox/pull/468/commits/d65e03cb622e0e74c844be4adc55a0a5fc386a71.